### ISSUE TITLE
`highlight_resolved_externals` slows down typing in `nushell` when computer is switched from AC to battery power

### ISSUE THREAD

**Issue by NoahMDMi** [nushell/nushell#13166]
### Describe the bug

I've noticed that while using nushell, if I unplug my laptop, typing in the shell becomes very slow and laggy. If I set `highlight_resolved_externals` to `false` this issue does not arise.

### How to reproduce

1. Start a nushell session while connected to AC power
2. Switch to battery power
3. Type in session

### Expected behavior

I expected nushell to remain as responsive as it was on AC power. 

### Configuration

| key                | value                                           |
| ------------------ | ----------------------------------------------- |
| version            | 0.94.2                                          |
| major              | 0                                               |
| minor              | 94                                              |
| patch              | 2                                               |
| branch             |                                                 |
| commit_hash        | be8c1dc0066cd1034a6b110a622f47b516bfe029        |
| build_os           | windows-x86_64                                  |
| build_target       | x86_64-pc-windows-msvc                          |
| rust_version       | rustc 1.77.2 (25ef9e3d8 2024-04-09)             |
| rust_channel       | 1.77.2-x86_64-pc-windows-msvc                   |
| cargo_version      | cargo 1.77.2 (e52e36006 2024-03-26)             |
| build_time         | 2024-06-03 19:27:48 +00:00                      |
| build_rust_channel | release                                         |
| allocator          | mimalloc                                        |
| features           | default, sqlite, system-clipboard, trash, which |
| installed_plugins  |                                                 |

### Additional context

I am running nushell via Windows Terminal (Preview). 

Additionally, I feel as if I remember that the `highlight_resolved_externals` is making a `which` call each time to resolve those external commands. If that is the case, I realize that this may not be something that can be fixed necessarily, but was wondering if there was any workarounds, for example:

1. A way to set the variable to false when running on battery power
2. Supplement the performance of `which` by caching resolved externals
3. Changing the behavior of how often `which` is run? (Seems to be checking the whole command on keystroke)
4. Allow the externals resolver to be defined manually (Would allow users to define caching manually, use different strategies?)

I am open to any suggestions! 

Thank you!

**Comment by fdncred** [comment]
This is why when I added it, I added an option to turn it off. You're right that this functionality does a `which` with each character typed. I thought about caching it but then it gets quite a bit more complicated because you have to know when to invalidate the cache and re-cache every command. You also have to know how much memory to use and limit it somehow because now you're holding a list of every accessible executable in memory all the time.

I don't think there is a nushell command that works cross-platform that tells you when you're on battery power or not. If you had one with your OS, you could setup an env_change hook or something like that to turn the config setting off.

**Comment by NoahMDMi** [comment]
Yeah, I figured as much. It's always tricky to determine those requirements. I will look into doing an `env_change` hook for the time being. 

Is there any way to override the `which` command that is getting called when that variable is set to `true`. I suppose there is a worry that someone might abuse that, but could be an opportunity for users to have their own custom logic?

**Comment by fdncred** [comment]
The only option right now is to set $env.config.highlight_resolved_externals to false to disable it or true to enable it.
